File: Cuarto/ProgramaciónDistribuida/python_concurrencia2/ejercicio2.py
import threading
import random

class Jugador(threading.Thread):
    def __init__(self, id, dado1, dado2):
        super().__init__()
        self.id = id
        self.dado1 = dado1
        self.dado2 = dado2
        self.puntos = 0
    
    def run(self):
        print(f"Jugador {self.id} ha lanzado los dados")
        dado1.lanzar()
        dado2.lanzar()
        self.puntos = dado1.valor + dado2.valor
        print(f"Jugador {self.id} ha sacado {self.puntos} puntos, con los dados {dado1.valor} y {dado2.valor}")
        dado1.liberar()
        dado2.liberar()
    
    def get_resultado(self):
        return self.puntos

class Dado():
    def __init__(self):
        self.disponible = True
        self.valor = 0
        self.cond = threading.Condition()
    
    def lanzar(self):
        self.cond.acquire()
        if (not self.disponible): self.cond.wait()
        self.disponible = False
        self.valor = random.randint(1,6)
        self.cond.release()
    
    def liberar(self):
        self.cond.acquire()
        self.disponible = True
        self.cond.notify()
        self.cond.release()
    
    def get_valor(self):
        return self.valor

dado1 = Dado()
dado2 = Dado()

thread_p1 = Jugador(1, dado1, dado2)
thread_p2 = Jugador(2, dado1, dado2)

thread_p1.start()
thread_p2.start()

thread_p1.join()
thread_p2.join()

if (thread_p1.get_resultado() > thread_p2.get_resultado()):
    print(f"Jugador 1 ha ganado con {thread_p1.get_resultado()} puntos")
elif (thread_p1.get_resultado() < thread_p2.get_resultado()):
    print(f"Jugador 2 ha ganado con {thread_p2.get_resultado()} puntos")
else:
    print(f"Empate con {thread_p1.get_resultado()} puntos")

File: Cuarto/ProgramaciónDistribuida/java_concurrencia2/ejercicio2.java
package java_concurrencia2;

import java.util.Random;

public class ejercicio2 {
    public static class Jugador extends Thread{
        private int resultado;
        private int id;

        public Jugador(int id) {
            this.id = id;
        }

        public void run() {
            dado1.lanzarDados();
            dado2.lanzarDados();
            System.out.println("Jugador " + id + " lanza los dados");
            System.out.println("Obtiene " + dado1.getResultado() + " y " + dado2.getResultado());
            resultado = dado1.getResultado() + dado2.getResultado();
            dado1.cederDado();
            dado2.cederDado();
        }

        public int getResultado() {
            return resultado;
        }
    }

    public static class Dados{
        private int resultado;
        private boolean disponible;
        
        public Dados() {
            this.disponible = true;
        }

        public synchronized void lanzarDados() {
            if (!disponible) {
                try {
                    wait();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
            disponible = false;
            Random rand = new Random();
            resultado = rand.nextInt(6) + 1;
        }

        public synchronized void cederDado(){
            disponible = true;
            notify();
        }

        public int getResultado() {
            return resultado;
        }
    }

    static Dados dado1 = new Dados();
    static Dados dado2 = new Dados();

    public static void main(String[] args) {
        Thread jugador1 = new Jugador(1);
        Thread jugador2 = new Jugador(2);

        jugador1.start();
        jugador2.start();

        try {
            jugador1.join();
            jugador2.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        
        if (((Jugador)jugador1).getResultado() > ((Jugador)jugador2).getResultado()) {
            System.out.println("Gana jugador 1");
        } else if (((Jugador)jugador1).getResultado() < ((Jugador)jugador2).getResultado()) {
            System.out.println("Gana jugador 2");
        } else {
            System.out.println("Empate");
        }
    }
}
